parser/src/ast/callable.rs

```diff
@@ -16,3 +20,60 @@ pub struct FunDeclaration<'a> {
     pub parameters: Vec<TypedVariable<'a>>,
     pub body: Vec<Expr<'a>>,
 }
+
+/// A redirection.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Redir<'a> {
+    /// File descriptor that is modified by this redirection.
+    pub fd: RedirFd,
+    /// The nature of the redirection.
+    pub operator: RedirOp,
+    /// The file name or file descriptor to redirect to.
+    pub operand: Expr<'a>,
+}
+
+/// A file descriptor that is redirected.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum RedirFd {
+    /// The default file descriptor, if not specified.
+    ///
+    /// This often means to use the standard input or output.
+    Default,
+    /// All the standard output file descriptors, grouped with `&`.
+    Wildcard,
+    /// A specific file descriptor.
+    Fd(u32),
+}
+
+/// Commands separated by `|`
+#[derive(Clone, Debug, PartialEq)]
+pub struct Pipeline<'a> {
+    /// Elements of the pipeline.
+    ///
+    /// A valid pipeline must have at least one command.
+    pub commands: Vec<Call<'a>>,
+    /// True if the pipeline begins with a `!`.
+    pub negation: bool,
```

Comment:
I don't think this should be at i the Call structure. As the `!` is an operator, it should be represented somewhere else in the AST imo